packages/react-scripts/scripts/utils/verifyTypeScriptSetup.js

```diff
@@ -250,6 +250,27 @@ function verifyTypeScriptSetup() {
     }
   }
 
+  if (parsedCompilerOptions.jsxImportSource != null) {
+    if (hasJsxRuntime && semver.gte(ts.version, '4.1.0-beta')) {
+      if (process.env.JSX_IMPORT_SOURCE == null) {
+        process.env.JSX_IMPORT_SOURCE = parsedCompilerOptions.jsxImportSource;
```

Comment:
IMHO this shouldn't set this global but rather it should be passed around using normal arguments etc. Assigning to a global pollutes the global env and might be surprising - the relation between the files involved in this becomes much more implicit and much less obvious.

Reply:
Thanks for taking a look at this @Andarist!
I know, I'm not a big fan either. I wrote it this way because an environment variable is used to control the related JSX transform (`DISABLE_NEW_JSX_TRANSFORM`), because it enables the feature for JavaScript users and because it made a first version easy to implement 🙂 

An alternative implementation, at least for the TypeScript use cases, would be extracting the parsing of the `tsconfig.json` file and passing the result to the verify function and (a part of it) as an extra argument to `configFactory`.

Based on feedback on this PR, I can change the implementation to one that the maintainers are happy with.

Reply:
The usage of `process.env.JSX_IMPORT_SOURCE` is not that bad - as you have mentioned, some other stuff is handled in a similar manner. The problem (from my PoV) is that you also *write* to a `process.env` and not only read from it.

Also - I'm just a bystander here, so it might be best to wait for maintainers' feedback before pursuing anything that I suggest 😉